tests/ui_tools/test_popups.py

```diff
@@ -187,6 +188,131 @@ def test_feature_level_content(self, mocker, zulip_version):
+                            return_value=self.user_data)
+
+        self.user_info_view = UserInfoView(self.controller, 10000,
+                                           'User Info (up/down scrolls)')
+
+    @pytest.mark.parametrize([
+        'initial_key_val',
+        'expected_key_val',
+    ], [
+        (None,                        ('Email', 'user@example.com')),
+        (('email', ''),               ('Email', None)),
+        (None,                        ('Date joined', '2021-03-18')),
+        (('date_joined', ''),         ('Date joined', None)),
+        (None,                        ('Timezone', 'Asia/Kolkata')),
+        (('timezone', ''),            ('Timezone', None)),
+        (None,                        ('Local time', '10:55 AM')),
+        (('local_time', ''),          ('Local time', None)),
+        (None,                        ('Last active', 'Tue Mar 13 10:55:22')),
+        (('last_active', ''),         ('Last active', None)),
+        (('bot_owner', 'Test Owner'), ('Owner', 'Test Owner')),
+        (None,                        ('Owner', None)),
+        (('bot_type', 1),             ('Role', 'Generic Bot')),
+        (('bot_type', 2),             ('Role', 'Incoming Webhook Bot')),
+        (('bot_type', 3),             ('Role', 'Outgoing Webhook Bot')),
+        (('bot_type', 4),             ('Role', 'Embedded Bot')),
+        (('role', [('is_owner', True),
+                   ('is_admin', False),
+                   ('is_guest', False),
+                   ]),                ('Role', 'Owner')),
+        (('role', [('is_owner', False),
+                   ('is_admin', True),
+                   ('is_guest', False),
+                   ]),                ('Role', 'Admin')),
+        (('role', [('is_owner', False),
+                   ('is_admin', False),
+                   ('is_guest', True),
+                   ]),                ('Role', 'Guest')),
+        (None,                        ('Role', 'Member')),
+    ], ids=[
+        'user_email',       'user_empty_email',
+        'user_date_joined', 'user_empty_date_joined',
+        'user_timezone',    'user_empty_timezone',
+        'user_local_time',  'user_empty_local_time',
+        'user_bot_owner',   'user_empty_bot_owner',
+        'user_last_active', 'user_empty_last_active',
+        'user_is_generic_bot',
+        'user_is_incoming_webhook_bot',
+        'user_is_outgoing_webhook_bot',
+        'user_is_embedded_bot',
+        'user_is_owner',
+        'user_is_admin',
+        'user_is_guest',
+        'user_is_member',
+    ])
+    def test__fetch_user_data(self, mocker, initial_key_val, expected_key_val):
+        data = self.user_data
+
+        if initial_key_val:
+            i_key, i_val = initial_key_val
+            data[i_key] = i_val
+            if i_key == 'bot_type':
+                data['is_bot'] = True
+                data['bot_owner'] = 'Test Owner'
+        e_key, e_val = expected_key_val
```

Comment:
If we used something like the `dict(fixture_copy, **to_vary_in_fixture)` approach as used in other tests, would this be clearer? (also your other similar tests)
(the bot parts can be part of the to_vary element?)